Source/DafnyStandardLibraries/src/Std/Parsers/Core/Parsers.dfy

```diff
@@ -0,0 +1,756 @@
+    // and will be propagated to the top ASAP
+    // A Recoverable error can typically be processed.
+    // Comittedness of the parser only depends if the .Remaining()
+    // of the parse result has moved since the input was provided.
+    Fatal | Recoverable
+
+  datatype ParseResult<+R> =
+      // ParseResult is the type of what a parser taking a seq<C> would return
+    | ParseFailure(level: FailureLevel, data: FailureData)
+      // Returned if a parser failed.
+    | ParseSuccess(result: R, remaining: Input)
+  // Returned if a parser succeeds, with the increment in the position
+  {
+    function Remaining(): Input
+      // If Remaining() is the same as the input, the parser is "uncommitted",
+      // which means combinators like Or and ZeroOrMore can try alternatives
+    {
+      if ParseSuccess? then remaining else data.remaining
+    }
+
+    predicate IsFailure() {
+      ParseFailure?
+    }
+
+    predicate IsFatalFailure() {
+      ParseFailure? && level == Fatal
+    }
+
+    predicate IsFatal()
+      requires IsFailure()
+    {
+      level == Fatal
+    }
+
+    function PropagateFailure<U>(): ParseResult<U>
+      requires IsFailure()
+    {
+      ParseFailure(level, data)
+    }
+
+    function Extract(): (R, Input)
+      requires !IsFailure()
+    {
+      (result, remaining)
+    }
+
+    function Map<R'>(f: R -> R'): ParseResult<R'>
+      // Transforms the result of a successful parse result
+    {
+      match this
+      case ParseSuccess(result, remaining) =>
+        ParseSuccess(f(result), remaining)
+      case ParseFailure(level, data) =>
+        ParseFailure(level, data)
+    }
+
+    function MapRecoverableError(
+      f: FailureData -> FailureData
+    ): ParseResult<R>
+      // If the result is a recoverable error,
+      // let the function process it
+    {
+      match this
+      case ParseFailure(Recoverable, data) =>
+        ParseFailure(Recoverable, f(data))
+      case _ => this
+    }
+
+    predicate NeedsAlternative(input: Input)
+      // Returns true if the parser result
+      // - Is a failure
+      // - Is recoverable
+      // - Did not consume any input (not-committed)
+    {
+      ParseFailure? && level == Recoverable && input == Remaining()
+    }
+  }
+
+  predicate IsRemaining(input: Input, remaining: Input)
+    // True if remaining is a suffix of the input
+  {
+    && A.Length(remaining) <= A.Length(input)
+    && A.Drop(input, A.Length(input)-A.Length(remaining)) == remaining
+  }
+
+  lemma IsRemainingTransitive(input: Input, remaining1: Input, remaining2: Input)
+    requires IsRemaining(input, remaining1)
+    requires IsRemaining(remaining1, remaining2)
+    ensures IsRemaining(input, remaining2)
+  {
+    A.AboutDrop(input, A.Length(input)-A.Length(remaining1), A.Length(remaining1)-A.Length(remaining2));
+  }
+
+  // Cannot express this predicate if Input is allocated. Add once we accept quantification over unallocated objects
```

Comment:
You might be interested in the "proof trait" pattern I used in the Actions library: https://github.com/dafny-lang/dafny/pull/6074/files#diff-cb5145d4b584ea7656360f05c946523c48c5651815cea269b469a855844c3ed2R105